vm/src/builtins/genericalias.rs

```diff
@@ -0,0 +1,166 @@
+use crate::builtins::{PyStr, PyTuple, PyTupleRef, PyTypeRef};
+use crate::common::hash;
+use crate::slots::{Hashable, SlotConstructor};
+use crate::{
+    IdProtocol, PyClassImpl, PyContext, PyObjectRef, PyRef, PyResult, PyValue, TryFromObject,
+    TypeProtocol, VirtualMachine,
+};
+use std::fmt;
+
+#[pyclass(module = "types", name = "GenericAlias")]
+pub struct PyGenericAlias {
+    origin: PyTypeRef,
+    args: PyTupleRef,
+    parameters: PyTupleRef,
+}
+
+pub type PyGenericAliasRef = PyRef<PyGenericAlias>;
+
+impl fmt::Debug for PyGenericAlias {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("GenericAlias")
+    }
+}
+
+impl PyValue for PyGenericAlias {
+    fn class(vm: &VirtualMachine) -> &PyTypeRef {
+        &vm.ctx.types.generic_alias_type
+    }
+}
+
+#[derive(FromArgs)]
+pub struct GenericAliasArgs {
+    #[pyarg(any)]
+    origin: PyTypeRef,
+    #[pyarg(any)]
+    arguments: PyObjectRef,
+}
+
+impl SlotConstructor for PyGenericAlias {
+    type Args = GenericAliasArgs;
+
+    fn py_new(cls: PyTypeRef, args: Self::Args, vm: &VirtualMachine) -> PyResult {
+        PyGenericAlias::new(args.origin, args.arguments, vm).into_pyresult_with_type(vm, cls)
+    }
+}
+
+#[pyimpl(with(Hashable), flags(BASETYPE))]
+impl PyGenericAlias {
+    pub fn new(origin: PyTypeRef, args: PyObjectRef, vm: &VirtualMachine) -> Self {
+        let args: PyTupleRef = if let Ok(tuple) = PyTupleRef::try_from_object(vm, args.clone()) {
+            tuple
+        } else {
+            vm.ctx
+                .new_tuple(vec![args])
+                .downcast::<PyTuple>()
+                .ok()
+                .unwrap()
+        };
+
+        Self {
+            origin,
+            args: args.clone(),
+            parameters: make_parameters(args, vm),
+        }
```

Comment:
because `make_parameters` doesn't consume args inside, `args.clone()` is redundant.

```suggestion
        let parameters = make_parameters(&args, vm);
        Self {
            origin,
            args,
            parameters,
        }
```